<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Structure Constraints • shine</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Structure Constraints">
<meta property="og:description" content="shine">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">shine</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Examples
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/docs/simulations.html">Simulating Graphs</a>
    </li>
    <li>
      <a href="../../articles/docs/filtering.html">Filtering Methods</a>
    </li>
    <li>
      <a href="../../articles/docs/constraints.html">Structure Constraints</a>
    </li>
    <li>
      <a href="../../articles/docs/learning.html">Network Learning</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Functionality</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/montilab/shine">
    <span class="fas fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="constraints_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Structure Constraints</h1>
            
      
      
      <div class="hidden name"><code>constraints.Rmd</code></div>

    </div>

    
    
<p>To aid in the inference of networks with small sample sizes, we apply biolgically sensible constraints to learning the network structure. To find constraints for the networks, we first establish a network structure in a lower dimensional space using zero-order correlation between genes, which can be computed with few samples.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">shine</span><span class="op">)</span></code></pre></div>
<div id="simulated-data" class="section level1">
<h1 class="hasAnchor">
<a href="#simulated-data" class="anchor"></a>Simulated Data</h1>
<p>As an example, we generate an undirected scale-free network using the Barabási–Albert model with 300 nodes (n) and simulate 100 samples (p) that follow a multivariate normal distribution.</p>
<div id="generate-scale-free-graph" class="section level3">
<h3 class="hasAnchor">
<a href="#generate-scale-free-graph" class="anchor"></a>Generate Scale-Free Graph</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"shine"</span><span class="op">)</span>, <span class="st">"models.rds"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../../reference/model.plot.html">model.plot</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></code></pre></div>
<p><img src="constraints_files/figure-html/unnamed-chunk-3-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div id="simulate-multivariate-gaussian-data" class="section level3">
<h3 class="hasAnchor">
<a href="#simulate-multivariate-gaussian-data" class="anchor"></a>Simulate Multivariate Gaussian Data</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/model.sim.html">model.sim</a></span><span class="op">(</span><span class="fl">100</span>, <span class="va">g</span>, seed<span class="op">=</span><span class="fl">1</span><span class="op">)</span>
<span class="va">bdg</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">bdg</span>
<span class="va">eset</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">eset</span></code></pre></div>
</div>
</div>
<div id="constraint-methods" class="section level1">
<h1 class="hasAnchor">
<a href="#constraint-methods" class="anchor"></a>Constraint Methods</h1>
<p>We use a specific version of the <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-9-559">Weighted Gene Co-Expression Analysis</a> method to identify gene modules where are clustered by their co-expression similarity and hierarchically clustered into gene modules of tightly co-expressed genes. However any method of identifying co-expression modules is applicable. Establishing gene modules captures coordinated co-expression of genes in a network that approximately follows a scale free topology.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">wgcna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/mods.detect.html">mods.detect</a></span><span class="op">(</span><span class="va">eset</span>, min.sft<span class="op">=</span><span class="fl">0.85</span>, beta<span class="op">=</span><span class="fl">3</span>, min.size<span class="op">=</span><span class="fl">10</span>, cores<span class="op">=</span><span class="fl">1</span>, do.plot<span class="op">=</span><span class="cn">T</span><span class="op">)</span></code></pre></div>
<p><img src="constraints_files/figure-html/unnamed-chunk-5-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">wgcna</span><span class="op">$</span><span class="va">colors</span><span class="op">)</span></code></pre></div>
<p>Here we relay the modules back on the true graphs structure…</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/V.html">V</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">$</span><span class="va">color</span> <span class="op">&lt;-</span> <span class="va">wgcna</span><span class="op">$</span><span class="va">colors</span>
<span class="fu"><a href="../../reference/model.plot.html">model.plot</a></span><span class="op">(</span><span class="va">g</span>, colors<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="constraints_files/figure-html/unnamed-chunk-6-1.png" width="700" style="display: block; margin: auto;"></p>
<div id="isolated-modules" class="section level2">
<h2 class="hasAnchor">
<a href="#isolated-modules" class="anchor"></a>Isolated Modules</h2>
<p>Now that we’ve identified co-expression modules, we could build networks for each module independently, assuming there is no cross-talk between module genes.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">wgcna</span><span class="op">$</span><span class="va">mods</span><span class="op">$</span><span class="va">grey</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">wgcna</span><span class="op">$</span><span class="va">mods</span><span class="op">)</span></code></pre></div>
<pre><code>List of 17
 $ pink        : chr [1:27] "1" "33" "36" "54" ...
 $ black       : chr [1:27] "2" "32" "69" "93" ...
 $ green       : chr [1:31] "3" "9" "48" "89" ...
 $ yellow      : chr [1:33] "5" "11" "13" "35" ...
 $ purple      : chr [1:26] "6" "18" "28" "94" ...
 $ salmon      : chr [1:22] "7" "12" "15" "39" ...
 $ red         : chr [1:29] "8" "55" "59" "63" ...
 $ greenyellow : chr [1:25] "14" "88" "119" "123" ...
 $ brown       : chr [1:35] "16" "19" "25" "27" ...
 $ turquoise   : chr [1:58] "17" "66" "84" "101" ...
 $ lightcyan   : chr [1:13] "20" "61" "152" "155" ...
 $ tan         : chr [1:24] "21" "34" "46" "131" ...
 $ magenta     : chr [1:26] "22" "29" "31" "49" ...
 $ grey60      : chr [1:13] "23" "43" "53" "65" ...
 $ midnightblue: chr [1:14] "24" "76" "104" "148" ...
 $ blue        : chr [1:43] "26" "50" "51" "58" ...
 $ cyan        : chr [1:19] "37" "41" "42" "77" ...</code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ig.iso</span> <span class="op">&lt;-</span> <span class="va">g</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">x</span> <span class="kw">in</span> <span class="va">wgcna</span><span class="op">$</span><span class="va">mods</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/combn.html">combn</a></span><span class="op">(</span><span class="va">x</span>, <span class="fl">2</span><span class="op">)</span>
    <span class="va">e</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/edge.html">edges</a></span><span class="op">(</span><span class="va">pairs</span><span class="op">)</span>
    <span class="va">ig.iso</span> <span class="op">&lt;-</span> <span class="va">ig.iso</span> <span class="op">+</span> <span class="va">e</span>
<span class="op">}</span>
<span class="fu"><a href="../../reference/model.plot.html">model.plot</a></span><span class="op">(</span><span class="va">ig.iso</span>, colors<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="constraints_files/figure-html/unnamed-chunk-7-1.png" width="700" style="display: block; margin: auto;"></p>
<p>While this would greatly reduce the dimensionality of the problem, this will result in a disconnected global structure with no inter-module connections! This should only be used in the most extreme cases where the search complexity is too high.</p>
</div>
<div id="interacting-modules" class="section level2">
<h2 class="hasAnchor">
<a href="#interacting-modules" class="anchor"></a>Interacting Modules</h2>
<p>Alternatively, we could limit interactions between modules by identifying modules likely to be interacting. In other words, we look for a meta network at the module level first. Using a meta network of co-regulated gene modules, we can represent each module as an eigengene, which is the first principle component (PC1) of the expression values of genes within a given module, resulting in a module x sample matrix. Within this lower-dimensional representation of the network, we establish conditional independence between modules by measuring the pairwise partial correlation.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">wgcna</span><span class="op">$</span><span class="va">mods</span><span class="op">$</span><span class="va">grey</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>    
<span class="va">wgcna</span><span class="op">$</span><span class="va">eigengenes</span><span class="op">$</span><span class="va">grey</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>
<span class="va">metanet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/bdg.metanet.html">bdg.metanet</a></span><span class="op">(</span><span class="va">wgcna</span><span class="op">$</span><span class="va">eigengenes</span>, mpl<span class="op">=</span><span class="cn">TRUE</span>, cut<span class="op">=</span><span class="fl">0.25</span>, iter<span class="op">=</span><span class="fl">25000</span>, cores<span class="op">=</span><span class="fl">3</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">V</span><span class="op">(</span><span class="va">metanet</span><span class="op">$</span><span class="va">ig</span><span class="op">)</span><span class="op">$</span><span class="va">color</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu">V</span><span class="op">(</span><span class="va">metanet</span><span class="op">$</span><span class="va">ig</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">metanet</span><span class="op">$</span><span class="va">ig</span>, vertex.size<span class="op">=</span><span class="fl">10</span>, edge.width<span class="op">=</span><span class="fl">3</span>, vertex.label<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></code></pre></div>
<p><img src="constraints_files/figure-html/unnamed-chunk-9-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Now we might enforce constraints where interactions are allowed within modules or between connected modules.</p>
</div>
<div id="extended-modules" class="section level2">
<h2 class="hasAnchor">
<a href="#extended-modules" class="anchor"></a>Extended Modules</h2>
<p>Alternatively to this all-or-nothing approach (when considering inter-module connections), we could determine secondary membership of individual nodes, which would result in inter-module interactions. We do this through the concept of fuzzy membership, where the membership of an individual node can be considered for each module.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/mod.plot.html">mod.plot</a></span><span class="op">(</span><span class="va">wgcna</span><span class="op">$</span><span class="va">dat</span>, <span class="va">wgcna</span><span class="op">$</span><span class="va">mods</span>, mod<span class="op">=</span><span class="st">"black"</span>, <span class="va">wgcna</span><span class="op">$</span><span class="va">colors</span>, size<span class="op">=</span><span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="constraints_files/figure-html/unnamed-chunk-10-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Here we plot each node by its membership scores (MM1 &amp; MM2) for the black module. Membership of a node in a given module is its correlation with the first and second eigengene (PC1 &amp; PC2). Here we see the nodes with primary membership in the black module have high membership scores, as expected. Now assigning secondary membership to nodes not originally put in the black module is treated as a classification problem using QDA.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/fuzzy.plot.html">fuzzy.plot</a></span><span class="op">(</span><span class="va">wgcna</span><span class="op">$</span><span class="va">dat</span>, <span class="va">wgcna</span><span class="op">$</span><span class="va">mods</span>, mod<span class="op">=</span><span class="st">"black"</span>, p<span class="op">=</span><span class="fl">0.9</span><span class="op">)</span></code></pre></div>
<p><img src="constraints_files/figure-html/unnamed-chunk-11-1.png" width="700" style="display: block; margin: auto;"></p>
<p>We extend the black module to its primary and now secondary members. And repeat the process for extending every module.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/mods.plot.html">mods.plot</a></span><span class="op">(</span><span class="va">wgcna</span><span class="op">$</span><span class="va">dat</span>, <span class="va">wgcna</span><span class="op">$</span><span class="va">mods</span>, <span class="va">wgcna</span><span class="op">$</span><span class="va">colors</span>, ncol<span class="op">=</span><span class="fl">5</span>, size<span class="op">=</span><span class="fl">2.5</span><span class="op">)</span></code></pre></div>
<p><img src="constraints_files/figure-html/unnamed-chunk-12-1.png" width="1152" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fuzzy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/fuzzy.mods.html">fuzzy.mods</a></span><span class="op">(</span><span class="va">wgcna</span><span class="op">$</span><span class="va">dat</span>, <span class="va">wgcna</span><span class="op">$</span><span class="va">mods</span>, p<span class="op">=</span><span class="fl">0.9</span><span class="op">)</span>
<span class="va">fuzzy</span><span class="op">$</span><span class="va">grey</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>
<span class="va">mods</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">wgcna</span><span class="op">$</span><span class="va">mods</span>, <span class="va">length</span><span class="op">)</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">fuzzy</span>, <span class="va">length</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">mods</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<pre><code>   turquoise         blue        brown       yellow        green          red 
          68           50           46           40           36           35 
        pink        black       purple      magenta  greenyellow          tan 
          32           36           31           32           31           29 
      salmon         cyan midnightblue    lightcyan       grey60 
          27           26           17           19           19 </code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">fuzzy</span>, <span class="va">length</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">mods</span><span class="op">)</span><span class="op">]</span><span class="op">-</span><span class="va">mods</span></code></pre></div>
<pre><code>   turquoise         blue        brown       yellow        green          red 
          10            7           11            7            5            6 
        pink        black       purple      magenta  greenyellow          tan 
           5            9            5            6            6            5 
      salmon         cyan midnightblue    lightcyan       grey60 
           5            7            3            6            6 </code></pre>
<p>With our modules extending, we enforce constraints where interactions are allowed within these extended or fuzzy modules. Thus intra-module interactions and only the most likely inter-module interactions will be considered.</p>
</div>
<div id="extensions" class="section level2">
<h2 class="hasAnchor">
<a href="#extensions" class="anchor"></a>Extensions</h2>
<p>This idea can be extended to the many module detection algorithms released in recent years. Check out <a href="https://www.nature.com/articles/s41467-018-03424-4">A Comprehensive Evaluation of Module Detection Methods for Gene Expression data</a> for an excellent review and comparison. Additionally you can check out these other <a href="https://github.com/anfederico/module-detection">module detection wrappers scripts</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/anfederico">Anthony Federico</a>, <a href="https://github.com/tetomonti">Stefano Monti</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
